### PR TITLE
Support slot tags,

### PR DESCRIPTION
**Pull request by tomalec**
translate them back to content

Fixes: Starcounter/starcounter-include/issues/19

I wonder, should I also use V1 Shadow DOM if available (`attachShadow`), and then translate content tags into slots - this may affect CSS selectors, and provide slightly more inconsistency between browsers.

We cna also consider, to implement it later in `declarative-shadow-dom`

### CODE REVIEW

File: starcounter-include.html [hunk 1]
```diff
@@ -288,6 +289,60 @@
...
         }
 
+        /**
+         * Replaces all <slot> tags with <content>,
+         *  preserves all attributes, translates selector to slot name
+         * @param  {String} string stringified HTML
+         * @return {String}        translated HTML
+         */
+        function replaceSlotsWithContentInString(string){
+            return string.replace(/<\/slot(\s*)>/gi,'</content$1>')
```
Comment (alshakero) [inline starcounter-include.html]: left side white space before tag name in the closing tag is valid HTML, I suggest adding another `\s*` to become `string.replace(/<\/\s*slot(\s*)>/gi,'</content$1>')`.

File: starcounter-include.html [hunk 2]
```diff
@@ -288,6 +289,60 @@
...
+            }
+            var name = slot.getAttribute('name');
+            if (name) {
+                content.setAttribute('select', '[slot=\'' + name + '\']');
+            }
+            slot.parentNode.replaceChild(content, slot);
+            return content;
+        }
+        /**
+         * Replace all <slot> elements with <content> elements i given document fragment
```
Comment (alshakero) [inline starcounter-include.html]: A little typo  `elements i given document`

File: starcounter-include.html [hunk 3]
```diff
@@ -288,6 +289,60 @@
...
+         * @param  {Documentfragment|Node} root scope. Will be changed.
+         * @return {Documentfragment|Node}      modified scope
+         */
+        function replaceSlotsWithContent(root) {
+            if (root.firstChild) {
+                var node = root.firstChild;
+                while (node) {
+                    var next = node.nextSibling;
```
Comment (alshakero) [inline starcounter-include.html:323]: Is `next` necessary? the created `content` node has a valid `nextSibling`. I think the following is sufficient
```
if (root.firstChild) {
    var node = root.firstChild;
    while (node) {
        if (node.localName == 'slot') {
            node = replaceSlotWithContent(node);
        }
        node = node.nextSibling;
    }
}
```
All tests pass with this code.

File: test/composition/slot-to-content.html [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+<head>
+    <meta charset="utf-8">
+    <meta name="viewport" content="width=device-width, minimum-scale=1.0, initial-scale=1.0, user-scalable=yes">
+
+    <script src="../../../webcomponentsjs/webcomponents.js"></script>
+    <script src="../../../web-component-tester/browser.js"></script>
+    <script src="../helpers/chainable-methods-for-partial.js"></script>
+
+    <!-- Step 1: import the element to test -->
```
Comment (alshakero) [inline test/composition/slot-to-content.html:12]: Couldn't find any other steps.

File: test/composition/slot-to-content.html [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    </template>
+
+    <script id="toBeLoadedAsString" type="text/html">
+        <h2>Custom Shadow DOM composition</h2>
+		<slot name="my-slot"></slot>
+		<slot name='my-slot'></slot>
+		<slot></slot>
+		<slot    name="my-slot"    ></slot>
+		<slot       ></slot   >
+		<slot foo="bar" name="my-slot" baz></slot>
```
Comment (alshakero) [inline test/composition/slot-to-content.html:45]: I think we need one more case for the comment above. _left side white space in closing tag_.

**Review by alshakero — COMMENTED** [review]
Well, the code is godly. Just simple suggestions.

File: starcounter-include.html (hunk 1 above)
Comment (tomalec) [reply starcounter-include.html]: I wouldn't bother about invalid HTML 
https://www.w3.org/TR/html5/syntax.html#end-tags

Comment (alshakero) [reply starcounter-include.html]: Well that's awesome, I checked with a two linters and with Chrome. And they both accepted it. But if W3C agrees with you, it's the linters' problem 💃

File: test/composition/slot-to-content.html (hunk 5 above)
Comment (alshakero) [reply test/composition/slot-to-content.html:45]: You can ignore this now.

File: starcounter-include.html (hunk 3 above)
Comment (tomalec) [reply starcounter-include.html:323]: It's for weird behaving custom elements in fallback content, that might mess around, plus I'm not sure how Shadow DOM dom polyfill distributes elements. Polymer does so in their shim.

Have you tested all the browsers?

File: test/composition/slot-to-content.html (hunk 4 above)
Comment (tomalec) [reply test/composition/slot-to-content.html:12]: Ha :) copy'n'paste leftover

File: starcounter-include.html (hunk 3 above)
Comment (alshakero) [reply starcounter-include.html:323]: Chrome ✔️ 
FF ✔️ 
Safari ✔️ 
IE ✔️ (one failing spec, but it fails with and without, no regression).

**Comment by warpech** [comment]
Note to self: ping @jwillmer when this is merged to Starcounter 2.3

**Comment by warpech** [comment]
This looks really good. Have you tested in a Starcounter app? Using a mix of V1 and V0, running in Chrome and shimmed browsers?

**Comment by tomalec** [comment]
It's not ready, I'm currently working on making this translation in `juicy-composition` this will be more convenient to do it closer to the shadow root itself, as we are setting it in multiple ways. Also it would be easier to get rid of it once we  will move to `declarative-shadow-dom` and V1

**Comment by warpech** [comment]
@tomalec right now we have a problem that apps that use `::content` but no `::slotted()` look wrong in Chrome (see People develop branch)

Will finishing this fix the issue?

**Comment by tomalec** [comment]
Could you open an issue and describe te problem in more details? This PR should be closed without merging, applicable changes are already merged and released via `juicy-composition`

**Pull request closed without merging**